File: code/python/disk_erase.py
import subprocess
import logging
import sys

logging.basicConfig(level=logging.INFO, format="%(asctime)s - %(levelname)s - %(message)s")

def is_ssd(device: str) -> bool:
    try:
        output = subprocess.run(
            ["cat", f"/sys/block/{device}/queue/rotational"],
            check=True,
            stdout=subprocess.PIPE,
            stderr=subprocess.PIPE,
        )
        return output.stdout.decode().strip() == "0"
    except FileNotFoundError:
        logging.error(f"SSD check failed: {device} not found.")
        sys.exit(2)
    except subprocess.SubprocessError:
        logging.error(f"Error executing subprocess for SSD check on {device}.")
        sys.exit(1)
    return False

def erase_disk(device: str, passes: int) -> None:
    try:
        # Type-casting arguments to ensure correct types
        device = str(device)
        passes = int(passes)

        if is_ssd(device):
            logging.info(f"Warning: {device} appears to be an SSD. Use `hdparm` for secure erase.")
            return

        logging.info(f"Erasing {device} using shred with {passes} passes...")

        for i in range(1, passes + 1):
            logging.info(f"Pass {i} of {passes} is being processed on disk {device}...")
            subprocess.run(["shred", "-n", "1", "-z", f"/dev/{device}"], check=True)

        logging.info(f"Wiping partition table of {device} using dd...")
        subprocess.run(["dd", "if=/dev/zero", f"of=/dev/{device}", "bs=1M", "count=10"], check=True)

        logging.info(f"Disk {device} successfully erased.")
    except FileNotFoundError:
        logging.error(f"Error: Required command not found.")
        sys.exit(2)
    except subprocess.CalledProcessError:
        logging.error(f"Error: Failed to erase {device} using shred or dd.")
        sys.exit(1)

File: code/c/src/disk_erase.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <unistd.h>
#include <errno.h>
#include "disk_erase.h"
#include "utils.h"

#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int is_ssd(const char *device) {
    char path[256];
    snprintf(path, sizeof(path), "/sys/block/%s/queue/rotational", device);

    FILE *file = fopen(path, "r");
    if (!file) {
        perror(path);
        return -1;
    }

    char buffer[10];
    if (!fgets(buffer, sizeof(buffer), file)) {
        perror("Error reading from file");
        fclose(file);
        return -1;
    }

    fclose(file);

    // Trim newline character
    buffer[strcspn(buffer, "\n")] = 0;

    // Convert to integer and check rotational value
    return atoi(buffer) == 0 ? 0 : 1;
}


void erase_disk(const char *device, int passes) {
    if (!is_ssd(device)) {
        printf("Warning: %s appears to be an SSD. Use `hdparm` for secure erase.\n", device);
        return;
    }

    printf("Erasing %s using shred with %d passes...\n", device, passes);

    for (int i = 1; i <= passes; i++) {
        printf("Pass %d of %d is being processed...\n", i, passes);
        char cmd[256];
        snprintf(cmd, sizeof(cmd), "shred -n 1 -z /dev/%s", device);
        if (system(cmd) != 0) {
            perror("Error executing shred command");
            exit(1);
        }
    }

    printf("Wiping partition table of %s using dd...\n", device);
    if (system("dd if=/dev/zero of=/dev/device bs=1M count=10") != 0) {
        perror("Error executing dd command");
        exit(1);
    }

    printf("Disk %s successfully erased.\n", device);
}
